catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/SolrSplitCatalogProvider.java

```diff
@@ -0,0 +1,652 @@
+
+  protected final List<SolrCollectionProvider> solrCollectionProviders;
+
+  protected final List<SolrCollectionListener> collectionListener;
+
+  private QueryResultSorter sorter = new QueryResultSorter();
+
+  private boolean getHandlerWorkaround = true;
+
+  private String collectionAlias = DEFAULT_QUERY_ALIAS;
+
+  /**
+   * Constructor that creates a new instance and allows for a custom {@link DynamicSchemaResolver}
+   *
+   * @param clientFactory Solr client factory
+   * @param adapter injected implementation of FilterAdapter
+   * @param resolver Solr schema resolver
+   */
+  public SolrSplitCatalogProvider(
+      SolrClientFactory clientFactory,
+      FilterAdapter adapter,
+      SolrFilterDelegateFactory solrFilterDelegateFactory,
+      DynamicSchemaResolver resolver,
+      List<SolrCollectionProvider> solrCollectionProviders,
+      List<SolrCollectionListener> collectionListener) {
+    Validate.notNull(clientFactory, "SolrClientFactory cannot be null.");
+    Validate.notNull(adapter, "FilterAdapter cannot be null");
+    Validate.notNull(solrFilterDelegateFactory, "SolrFilterDelegateFactory cannot be null");
+    Validate.notNull(resolver, "DynamicSchemaResolver cannot be null");
+    Validate.notEmpty(solrCollectionProviders, "CollectionProviders cannot be empty");
+    Validate.notNull(collectionListener, "SolrCollectionListener list must exist");
+
+    this.clientFactory = clientFactory;
+    this.filterAdapter = adapter;
+    this.solrFilterDelegateFactory = solrFilterDelegateFactory;
+    this.resolver = resolver;
+    this.solrCollectionProviders = solrCollectionProviders;
+    this.collectionListener = collectionListener;
+
+    initProviders();
+  }
+
+  /**
+   * Convenience constructor that creates a new ddf.catalog.source.solr.DynamicSchemaResolver
+   *
+   * @param clientFactory Solr client factory
+   * @param adapter injected implementation of FilterAdapter
+   */
+  public SolrSplitCatalogProvider(
+      SolrClientFactory clientFactory,
+      FilterAdapter adapter,
+      SolrFilterDelegateFactory solrFilterDelegateFactory,
+      List<SolrCollectionProvider> solrCollectionProviders,
+      List<SolrCollectionListener> collectionListener) {
+    this(
+        clientFactory,
+        adapter,
+        solrFilterDelegateFactory,
+        new DynamicSchemaResolver(),
+        solrCollectionProviders,
+        collectionListener);
+  }
+
+  protected BaseSolrCatalogProvider newProvider(String core) {
+    if (clientFactory.isSolrCloud() && core.equals(collectionAlias)) {
+      ensureAliasExists(getCollectionAlias(), getCatalogPrefix());
+    }
+    // TODO: using existing newClient method, could be simplified. Much of the logic is redundant,
+    // as configuration and solr remote collection is already created prior to this call.
+    return new BaseSolrCatalogProvider(
+        clientFactory.newClient(core), filterAdapter, solrFilterDelegateFactory, resolver);
```

Comment:
❓ What would need to change in DDF such that we could avoid this and compose the new components we needed downstream?

Reply:
avoid which part? the TODO comment?

Reply:
Yes, avoiding the redundant logic.

Reply:
So the current design choice might not be the most optimal but the least invasive to existing code. I am trying to take a two steps approach: first to introduce this split catalog provider using the same pattern as existing one. With minimal redesign to existing provider. Get that in first, since it is such a large code base, the longer we wait the more risk of divergent. Second step will refactor all together those two providers and various clean up, such as removing deprecated HttpSolrClientFactory, etc. Perhaps even including moving the entire solr related code to its own repo.

Reply:
circling back. this class is now renamed to SolrCatalogProviderImpl, the SplitSolrCatalogProvider actually compose SolrCatalogProviderImpl. The TODO is also addressed to overload newClient with additional properties